### DISCUSSION TITLE
How to set repl globals from code files?

### DISCUSSION THREAD

**Discussion by victorallume** [micropython/micropython#15103] [Using MicroPython]
I would like to, when the running program breaks into a repl (either through an exception or Ctrl-C), to set some variables to be accessible in the REPL. I've tried catching the KeyboardInterrupt, setting `globals()['foo'] = foo` then re-raising, but while that changes `globals()` with the exception handler, it doesn't pass it through to the repl. Any suggestions?

**Comment by Josverl** [comment]
How do you run your program?

- `import myprogram`
- `from my program import *`
- `exec(read(... ))`

Or perhaps yet another way?

**Reply by victorallume** [reply]
`mpconfigboard.h` has `#define MICROPY_BOARD_FROZEN_BOOT_FILE  "_boot.py"`
`_boot.py` is frozen, and calls `pyb.main('_app.py')`
`_app.py` is also frozen, and has:
```
from __app__ import main as app
await app()
```
`__app__.py` is on the filesystem, and is the main program. Each step does a bunch of other stuff, but this is the path to the main program.

**Comment by shariltumin** [comment]
I hope I understand your question correctly. Whenever I need to access a module variable, I just prepend the module name to the variable, for example:

```py
# hi.py
hi = 'hello'
name = 'Ali'

def hello():
    print(f'{hi} {name}')
```

Then I will be able to do this:
```bash
>>> import hi
>>> hi.hello()
hello Ali
>>> hi.hi='Hello'
>>> hi.name='Baba'
>>> hi.hello()
Hello Baba
>>> hi.what = 123.45
>>> dir(hi)
['__class__', '__name__', '__dict__', '__file__', 'name', 'hi', 'hello', 'what']
>>> print(hi.what*2)
246.9
```

I hope this helps to answer what you asked.

**Comment by andrewleech** [comment]
This can very much be done with aiorepl, with that package you can pass a custom global object to the repl when you start it. I often preload my aiorepl global with other imports and variables so they're ready to use without manually importing every time.

Aiorepl is then always available, even while your application is running. You don't need to break / interrupt the application to get to it. 

I'm not sure if this is possible at all from the base / startup repl.

**Reply by victorallume** [reply]
Thanks; we plan to eventually integrate aiorepl into our application, but aren't there quite yet. Looking for a stop-gap in the meantime.
